packages/client-http/src/client/Configuration.ts

```diff
@@ -8,18 +8,72 @@ export namespace Configuration {
     NoTreatmentMatch = 'RESOLVE_REASON_NO_TREATMENT_MATCH',
     Archived = 'RESOLVE_REASON_FLAG_ARCHIVED',
   }
+
+  export type FlagSchema =
+    | 'number'
+    | 'boolean'
+    | 'string'
+    | {
+        [step: string]: FlagSchema;
+      };
+
   export interface FlagValue<T = unknown> {
-    readonly value: T;
-    match<S>(obj: S): this is FlagValue<S>;
+    value: T;
+    schema: FlagSchema;
+  }
+  export interface Flag<T = unknown> extends FlagValue<T> {
+    name: string;
+    reason: ResolveReason;
+    variant: string;
+    value: T;
+    schema: FlagSchema;
+  }
+
+  export namespace FlagValue {
+    export function matches<T>({ schema }: FlagValue<T>, value: any): value is T {
```

Comment:
Should this be the other way around? We use the type of value to infer T? I have an example of that locally if you want to sync.